bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py

```diff
@@ -0,0 +1,512 @@
+        vectors = func(rdf)
+        v_results, v_res_types, v_nodes = vectors
+
+        # Convert the vector of results into a list so that we can mix
+        # different types in it.
+        # The std::tuple `vectors` is passed as parameter to tie its life with
+        # that of the new list
+        results = self.ListWithLifeline(v_results, vectors)
+
+        # Strip out the ROOT::RDF::RResultPtr<> part of the type
+        def get_result_type(s):
+            s = str(s)
+            pos = s.find('<')
+            if pos == -1:
+                raise RuntimeError(
+                    'Error parsing the result types of RDataFrame workflow')
+            return s[pos+1:-1].strip()
+
+        res_types = [ get_result_type(elem) for elem in v_res_types ]
+
+        # Add Python-only actions on their corresponding nodes
+        for (i, operation), n in zip(self.py_actions, v_nodes):
+            operation.kwargs['lazy'] = True  # make it lazy
+            res = getattr(n, operation.name)(*operation.args, **operation.kwargs)
+            results.insert(i, res)
+            res_types.insert(i, None) # placeholder
+
+        if v_results:
+            # We trigger the event loop here, so make sure we release the GIL
+            old_rg = RunGraphs.__release_gil__
+            RunGraphs.__release_gil__ = True
+            RunGraphs(v_results)
+            RunGraphs.__release_gil__ = old_rg
+
+        # Replace the RResultHandle of each Snapshot by its modified output
+        # path, since the latter is what we actually need in the reducer
+        for i, path in self.snapshots:
+            results[i] = [path]
+            res_types[i] = None # placeholder
+
+        # Replace the future-like result of every Python-only action (e.g.
+        # AsNumpyResult) by its actual value
+        for i, operation in self.py_actions:
+            results[i] = results[i].GetValue()
+
+        return results, res_types
+
+    def __repr__(self):
+        '''
+        Generates a string representation for this C++ workflow.
+
+        Returns:
+            string: code of this C++ workflow.
+        '''
+
+        return self._get_code()
+
+    def _get_code(self):
+        '''
+        Composes the workflow generation code from the different attributes
+        of this class. The resulting code contains a function that will be
+        called to generate the RDataFrame graph. Such function returns a tuple
+        of three elements:
+        1. A vector of results of the graph actions.
+        2. A vector with the result types of those actions.
+        3. A vector of RDF nodes that will be used in Python to invoke
+        Python-only actions on them (e.g. `AsNumpy`).
+        '''
+
+        code = '''
+{includes}
+
+namespace {namespace} {{
+
+typedef std::tuple<std::vector<ROOT::RDF::RResultHandle>,
+                   std::vector<std::string>,
+                   std::vector<ROOT::RDF::RNode>>
+        CppWorkflowResult;
```

Comment:
What's the advantage of one over the other here?

Reply:
no practical advantage here.

`using` is the superior alternative so that's what's used by default in modern C++ ("superior" means mostly just that it supports template type aliases).

(interestingly, as a consequence `typedef` often sends the message to the reader that they are reading old code)